aiidalab_widgets_base/structures.py

```diff
@@ -1,184 +1,206 @@
-        self.frozen = True
-        self.btn_store.disabled = True
-        self.structure_description.disabled = True
-        self.data_format.disabled = True
-
-    @property
-    def node_class(self):
-        return self.data_format.value
-
-    @node_class.setter
-    def node_class(self, value):
-        if self.frozen:
+        self.output.value = "Stored in AiiDA [{}]".format(self.structure_node)
+
+    @staticmethod
+    @default('node_class')
+    def _default_node_class():
+        return 'StructureData'
+
+    @observe('node_class')
+    def _change_structure_node(self, _=None):
+        self._structure_changed()
+
+    @default('structure')
+    def _default_structure(self):
+        return None
+
+    @validate('structure')
+    def _valid_structure(self, change):
+        """Returns ASE atoms object and sets structure_node trait."""
+
+        self._inserted_structure = change['value']
+
+        # If structure trait is set to Atoms object, structure node must be generated
+        if isinstance(self._inserted_structure, Atoms):
+            return self._inserted_structure
+
+        # If structure trait is set to AiiDA node, converting it to ASE Atoms object
+        if isinstance(self._inserted_structure, Data):
+            return self._inserted_structure.get_ase()
+
+        return None
+
+    @observe('structure')
+    def _structure_changed(self, _=None):
+        """This function enables/disables `btn_store` widget if structure is provided/set to None.
+        Also, the function sets `structure_node` trait to the selected node type."""
+
+        # If structure trait was set to None, structure_node should become None as well.
+        if self.structure is None:
+            self.set_trait('structure_node', None)
+            self.btn_store.disabled = True
             return
-        self.data_format.value = value
-
-    @property
-    def structure_node(self):
-        """Returns AiiDA StructureData node."""
-        if self._structure_node is None:
-            if self.structure_ase is None:
-                return None
-            # perform conversion
-            if self.data_format.value == 'CifData':
-                from aiida.orm.nodes.data.cif import CifData
-                self._structure_node = CifData()
-                self._structure_node.set_ase(self.structure_ase)
-            else:  # Target format is StructureData
-                self._structure_node = StructureData(ase=self.structure_ase)
-            self._structure_node.description = self.structure_description.value
-            self._structure_node.label = self.structure_ase.get_chemical_formula()
-        return self._structure_node
+
+        self.btn_store.disabled = False
+
+        # Chosing type for the structure node.
+        StructureNode = DataFactory(self.SUPPORTED_DATA_FORMATS[self.node_class])  # pylint: disable=invalid-name
+
+        # If structure trait is set to Atoms object, structure node must be created from it.
```

Comment:
I always thought that these kind of conversions are supposed to be handled in the "validate" functions, see the API documentation: https://traitlets.readthedocs.io/en/stable/defining_traits.html#traitlets.MyTrait.validate

> Check whether a given value is valid. If it is, it should return the value **(coerced to the desired type, if necessary)**[emphasis mine].

Either way is fine with me.

Reply:
Sure, if I would set `structure` trait, I would do it in `validate` function. However, here I am setting `structure_node` trait value.